README.md
## LDAP Authentication for DMX

### Configuration / Usage

In `config.properties` operatores have the following properties to configure:

#### LDAP Protocol without encryption

```
dmx.ldap.protocol = LDAP
dmx.ldap.server = 127.0.0.1
dmx.ldap.port = 389  
dmx.ldap.manager = 
dmx.ldap.password = 
dmx.ldap.user_base = 
dmx.ldap.user_attribute = 
dmx.ldap.user_filter = 
dmx.ldap.user_member_group = 
dmx.ldap.logging = INFO
```

Known Protocols are: 

- StartTLS (ldap://HOST:389) - default port is 389
- LDAPS (ldaps://HOST:636) - default port is 636
- LDAP (ldap://HOST:389) - default port is 389

Additional settings for self-signed certificates used with keystore:

```
javax.net.ssl.trustStore = /path/to/keystore.jks
javax.net.ssl.trustStorePassword = changeit
```

Note: For self signed certificates, dmx.ldap.server must contain the hostname of the certificate, not the IP address.

Known logging values are:
 
- INFO (default): Only warnings and errors are logged including possible misconfigurations.
- DEBUG: Hints, warning and errors are extensively logged during configuration and runtime phase.



CHANGELOG

Next Feature Release: 

* Fetch and write user profile data in LDAP System
* Creating new user accounts in LDAP System

**0.4.0** -- Apr 11, 2019

* Source code compatible with DMX
* Log messages contain configuration keys
* Fixed plugin crashing when no or no valid configuration given

**0.3.4** -- Nov 26, 2018

* Throw exceptions in cases of misconfiguration or any kind of error
* User creation, including automatic placing in memberOf group)
* Password change (SSHA passwords-only)
* Info and Debug Log setting
* Added many more logging locations
* Extensive checking of configuration validity (including keystore)

**0.3.0** -- Feb 27, 2018

* Add StartTLS and LDAPS/SSL protocols

**0.2.0** -- Feb 4, 2018

* Creates only a DM Username topic (along with private workspace) on first successful LDAP login
* Requires DeepaMehta 4.9.1

**0.1.0** -- Nov 30, 2017

* Basic functionality
* Creates full DM User Account on first successful LDAP login
* Requires DeepaMehta 4.9


src/main/java/de/deepamehta/ldap/Program.java
package de.deepamehta.ldap;

import java.util.Hashtable;

import javax.naming.Context;
import javax.naming.NamingException;
import javax.naming.directory.Attribute;
import javax.naming.directory.BasicAttribute;
import javax.naming.directory.BasicAttributes;
import javax.naming.directory.DirContext;
import javax.naming.directory.ModificationItem;
import javax.naming.ldap.Control;
import javax.naming.ldap.InitialLdapContext;
import javax.naming.ldap.LdapContext;

public class Program {

	public static void main(String[] args) throws Exception {
		System.out.println("user: " + args[0]);
		System.out.println("pass: " + args[1]);
		
		System.out.println("success: " + createUser(args[0], args[1]));
	}

	private static boolean createUser(String userName, String userPassword) {
		return createUserTest(
				"",
				"127.0.0.1",
				"389",
				"cn=admin,dc=dev,dc=climbo,dc=com",
				/*
				"cn=deepamehta,ou=systems,dc=dev,dc=climbo,dc=com",
				*/
				"12345678",
				"ou=users,dc=dev,dc=climbo,dc=com",
				"uid",
				userName,
				userPassword,
				"cn=climbousers,ou=groups,dc=dev,dc=climbo,dc=com");
	}
	
	static boolean createUserTest(
    		String protocol,
    		String server,
    		String port,
    		String manager,
    		String managerPassword,
    		String userBase,
    		String userAttribute,
    		String userName,
    		String userPassword,
    		String memberGroup) {
    	try {
    		LdapContext ctx = createContext(protocol, server, port, manager, managerPassword);
    		if (ctx == null) {
    			return false;
    		}
    		
    		return createUserImpl(ctx, userBase, userAttribute, userName, userPassword, memberGroup);
    				
    	} catch (Exception e) {
    		throw new RuntimeException("Creating user in LDAP failed", e);
    	}
    	
    }
	
    private static LdapContext createContext(
    		String ldapProtocol,
    		String ldapServer,
    		String ldapPort,
    		String ldapManager,
    		String ldapPassword) throws NamingException {
        final String port = (ldapPort == null) ? (ldapProtocol.equals("LDAPS") ? "636" : "389") : ldapPort;
        final String protocol = ldapProtocol.equals("LDAPS") ? "ldaps://" : "ldap://";
        final String server = protocol + ldapServer + ":" + port;
        
        return connect(server, ldapManager, ldapPassword);
    }

    private static LdapContext connect(String server, String username, String password) throws NamingException {
        Hashtable<String, Object> env = new Hashtable<String, Object>();
        env.put(Context.SECURITY_AUTHENTICATION, "simple");
        env.put(Context.PROVIDER_URL, server);
        env.put(Context.SECURITY_PRINCIPAL, username);
        env.put(Context.SECURITY_CREDENTIALS, password);
        env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory");

        //ensures that objectSID attribute values
        //will be returned as a byte[] instead of a String
        env.put("java.naming.ldap.attributes.binary", "objectSID");
        
        // the following is helpful in debugging errors
        // env.put("com.sun.jndi.ldap.trace.ber", System.err);
        Control[] arr = new Control[0];
        LdapContext ctx = new InitialLdapContext(env, arr);
        
        return ctx;
    }
    
    private static boolean createUserImpl(
    		LdapContext ctx,
    		String ldapUserBase,
    		String ldapUserAttribute,
    		String userName, String password,
    		String memberGroup) throws NamingException {
    	String entryDN = String.format("%s=%s,%s", ldapUserAttribute, userName, ldapUserBase);
    	
    	Attribute cn = new BasicAttribute("cn", userName);
    	Attribute sn = new BasicAttribute("sn", "deepamehta-ldap");
    	Attribute userPassword = new BasicAttribute("userPassword", password);

    	Attribute oc = new BasicAttribute("objectClass");
    	oc.add("top");
    	oc.add("person");
    	oc.add("organizationalPerson");
    	oc.add("inetOrgPerson");
    	
    	BasicAttributes entry = new BasicAttributes();
    	entry.put(oc);
    	entry.put(cn);
    	entry.put(sn);
    	entry.put(userPassword);
    	
    	ctx.createSubcontext(entryDN, entry);
    	
    	if (!memberGroup.isEmpty()) {
    		ModificationItem mi = new ModificationItem(DirContext.ADD_ATTRIBUTE,
    				new BasicAttribute("member", entryDN));
    		
    		try {
    			ctx.modifyAttributes(memberGroup, new ModificationItem[] { mi });
    		} catch (NamingException ne) {
    			// Removes user
    			ctx.destroySubcontext(entryDN);
    			
    			return false;
    		}
    	}
    	
    	return true;
    }

}
